src/dmd/root/array.d

```diff
@@ -591,3 +591,68 @@ unittest
     static assert (!is(typeof(Array!(OtherStruct).init.sort())));
     static assert (!is(typeof(Array!(OtherStruct).init.sort!pred)));
 }
+
+/**
+ * Filters the given array based on the given predicate.
+ *
+ * Returns: a new array containing only elements for which the predicate returns
+ *  `true`
+ */
+inout(T)[] filter(alias predicate, T)(inout T[] array)
+if (is(ReturnType!(typeof((T t) => predicate(t))) == bool))
+{
+    inout(T)[] result;
+
+    foreach (e ; array)
+    {
+        if (predicate(e))
+            result ~= e;
+    }
+
+    return result;
+}
+
+///
+pure nothrow @safe unittest
+{
+    const result = [1, 2, 3, 4].filter!(e => e > 2);
+    assert(result == [3, 4]);
+}
+
+/**
+ * Iterates the array and calls the given callable for each element.
+ *
+ * Returns: a new array containing the result of each call to `callable`
+ */
+auto map(alias callable, T)(inout T[] array) pure nothrow @safe
+if (__traits(compiles, { auto _ = (T t) => callable(t); } ))
+{
+    if (array.length == 0)
+        return null;
+
+    alias NewElementType = ReturnType!(typeof((T t) => callable(t)));
+
+    inout(NewElementType)[] result;
+    result.reserve(array.length);
```

Comment:
Assign to length and it'll work at ctfe as well;
and you don't have to use the concat operator further down

Reply:
`reserve` does work at CTFE. And I think it's an advantage to use concatenation instead of indexing.